File: leetcode/queue_stack/3221_Maximum_Array_Hopping_Score_II/codes/stack.py
from typing import List

class Solution:
    def maxScore(self, nums: List[int]) -> int:
        stack = [(nums[-1], len(nums) - 1)]
        for i in range(len(nums) - 2, -1, -1):
            if nums[i] > stack[-1][0]:
                stack.append((nums[i], i))
        res, j = 0, 0
        for num, ind in reversed(stack):
            res += (ind - j) * num
            j = ind
        return res

File: leetcode/queue_stack/3221_Maximum_Array_Hopping_Score_II/codes/stack.c
#include <stdio.h>

typedef long long ll;

struct Pair {
    int val;
    int idx;
};

ll maxScore(int* nums, int numsSize) {
    struct Pair stack[100005];
    int top = 0;
    stack[top++] = (struct Pair){ nums[numsSize - 1], numsSize - 1 };
    for (int i = numsSize - 2; i >= 0; --i) {
        if (nums[i] > stack[top - 1].val)
            stack[top++] = (struct Pair){ nums[i], i };
    }
    ll res = 0;
    int j = 0;
    for (int k = top - 1; k >= 0; --k) {
        res += (ll)(stack[k].idx - j) * stack[k].val;
        j = stack[k].idx;
    }
    return res;
}
